### PR TITLE
Don't interpret Period identifiers. Fixes #1058

### LINKED ISSUE #1058: Unable to process Amazon WorkMail server timezone data
**Describe the bug**
Calling `get_std_and_dst` on a `TimeZoneDefinition` obtained from Amazon WorkMail will cause an `IndexException` to be thrown.

**To Reproduce**
```python
account = # Create Amazon WorkMail based account
for zone in account.protocol.get_timezones([account.default_timezone], return_full_timezone_data=True):
    print(zone.get_std_and_dst(2022))
```

**Expected behavior**
```python
(StandardTime(bias=0, time=datetime.time(3, 0), occurrence=5, iso_month=10, weekday='Sunday'), DaylightTime(bias=-60, time=datetime.time(2, 0), occurrence=5, iso_month=3, weekday='Sunday'), Period(id='trule:Microsoft/Registry/W. Europe Standard Time/1-Standard', name='Standard', bias=datetime.timedelta(days=-1, seconds=82800)))
```

**Log output**
```python
Traceback (most recent call last):
  File "c:\Users\Mark\git\pyews\pyews.py", line 52, in <module>
    print(zone.get_std_and_dst(2022))
  File "C:\Users\Mark\venv\exchangelib\lib\site-packages\exchangelib\properties.py", line 1942, in get_std_and_dst
    standard_period = self._get_standard_period(for_year)
  File "C:\Users\Mark\venv\exchangelib\lib\site-packages\exchangelib\properties.py", line 1912, in _get_standard_period
    for period in sorted(self.periods, key=lambda p: (p.year, p.type)):
  File "C:\Users\Mark\venv\exchangelib\lib\site-packages\exchangelib\properties.py", line 1912, in <lambda>
    for period in sorted(self.periods, key=lambda p: (p.year, p.type)):
  File "C:\Users\Mark\venv\exchangelib\lib\site-packages\exchangelib\properties.py", line 1873, in year
    return self._split_id()[0]
  File "C:\Users\Mark\venv\exchangelib\lib\site-packages\exchangelib\properties.py", line 1868, in _split_id
    to_year, to_type = self.id.rsplit("/", 1)[1].split("-")
IndexError: list index out of range
```

**Additional context**
The `GetServerTimeZones` response from Amazon WorkMail:
```xml
<?xml version='1.0' encoding='utf-8'?>
<soap:Envelope
    xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns:xsd="http://www.w3.org/2001/XMLSchema">
  <Header xmlns="http://schemas.xmlsoap.org/soap/envelope/">
    <ServerVersionInfo xmlns="http://schemas.microsoft.com/exchange/services/2006/types" MajorVersion="14" MinorVersion="2" MajorBuildNumber="390" MinorBuildNumber="3" Version="Exchange2010_SP2"/>
  </Header>
  <soap:Body>
    <m:GetServerTimeZonesResponse
    xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types"
    xmlns:m="http://schemas.microsoft.com/exchange/services/2006/messages">
      <m:ResponseMessages>
        <m:GetServerTimeZonesResponseMessage ResponseClass="Success">
          <m:ResponseCode>NoError</m:ResponseCode>
          <m:TimeZoneDefinitions>
            <t:TimeZoneDefinition Id="W. Europe Standard Time" Name="(UTC+01:00) Amsterdam, Berlin, Bern, Rome, Stockholm, Vienna">
              <t:Periods>
                <t:Period Bias="-PT60M" Name="Standard" Id="std"/>
                <t:Period Bias="-PT120M" Name="Daylight" Id="dlt"/>
              </t:Periods>
              <t:TransitionsGroups>
                <t:TransitionsGroup Id="0">
                  <t:RecurringDayTransition>
                    <t:To Kind="Period">std</t:To>
                    <t:TimeOffset>PT180M</t:TimeOffset>
                    <t:Month>10</t:Month>
                    <t:DayOfWeek>Sunday</t:DayOfWeek>
                    <t:Occurrence>-1</t:Occurrence>
                  </t:RecurringDayTransition>
                  <t:RecurringDayTransition>
                    <t:To Kind="Period">dlt</t:To>
                    <t:TimeOffset>PT120M</t:TimeOffset>
                    <t:Month>3</t:Month>
                    <t:DayOfWeek>Sunday</t:DayOfWeek>
                    <t:Occurrence>-1</t:Occurrence>
                  </t:RecurringDayTransition>
                </t:TransitionsGroup>
              </t:TransitionsGroups>
              <t:Transitions>
                <t:Transition>
                  <t:To Kind="Group">0</t:To>
                </t:Transition>
              </t:Transitions>
            </t:TimeZoneDefinition>
          </m:TimeZoneDefinitions>
        </m:GetServerTimeZonesResponseMessage>
      </m:ResponseMessages>
    </m:GetServerTimeZonesResponse>
  </soap:Body>
</soap:Envelope>
```

### PR DESCRIPTION
**Pull request by cygnus9**
Instead find the correct period by finding the first Standard Period
referenced from the appropriate TransitionsGroup for the requested year.

### CODE REVIEW

File: exchangelib/properties.py [hunk 1]
```diff
@@ -1939,7 +1924,9 @@ def get_std_and_dst(self, for_year):
         if not 0 <= len(transitions_group.transitions) <= 2:
             raise ValueError(f"Expected 0-2 transitions in transitions group {transitions_group}")
 
-        standard_period = self._get_standard_period(for_year)
+        standard_period = self._get_standard_period(transitions_group)
```
Comment (ecederstrand) [inline exchangelib/properties.py:1927]: I'd prefer to keep the original contract where `_get_standard_period` throws the exception. I'm not very fond of communicating failure via `None`.

File: exchangelib/properties.py [hunk 2]
```diff
@@ -1906,17 +1894,14 @@ class TimeZoneDefinition(EWSElement):
...
 
-    def _get_standard_period(self, for_year):
-        # Look through periods and pick a relevant period according to the 'for_year' value
+    def _get_standard_period(self, transitions_group):
+        # Find the first standard period referenced from transitions_group
         valid_period = None
-        for period in sorted(self.periods, key=lambda p: (p.year, p.type)):
-            if period.year > for_year:
+        standard_periods_map = {p.id: p for p in self.periods if p.name == "Standard"}
+        for transition in transitions_group.transitions:
```
Comment (ecederstrand) [inline exchangelib/properties.py:1901]: Suggestion for a simpler logic:

```python
    def _get_standard_period(self, transitions_group):
        # Find the first standard period referenced from transitions_group
        standard_periods_map = {p.id: p for p in self.periods if p.name == "Standard"}
        for transition in transitions_group.transitions:
            try:
                return standard_periods_map[transition.to]
            except KeyError:
                continue
        raise TimezoneDefinitionInvalidForYear(f"Year {for_year} not included in periods {self.periods}")
```

**Comment by ecederstrand** [comment]
Thanks for the contribution!

This needs a matching test case. My suggestion would be to take the XML string from your issue and use that as input to the test case. Here's an example of doing that: https://github.com/ecederstrand/exchangelib/blob/bd4e67d8cfe5fce4c22e65607b8dc7a7a8483b80/tests/test_protocol.py#L369

That file would also be a suitable place to add the new test case.The relevant service to use for the test case is `exchangelib.services.GetServerTimeZones`

File: exchangelib/properties.py (hunk 1 above)
Comment (cygnus9) [reply exchangelib/properties.py:1927]: I totally agree. THe problem is that I don't have the `for_year` anymore in `_get_standard_period`. And it feels a bit artificial to pass it just for the exception. If you think that's fine I won't object to that though.

We can also just bubble up the `KeyError` exception, as described in your other suggestion, and deal with that at the call site? It does change the contract but at least failure is communicated explicitly.

File: exchangelib/properties.py (hunk 2 above)
Comment (cygnus9) [reply exchangelib/properties.py:1901]: What about:
``` python
    def _get_standard_period(self, transitions_group, for_year):
        # Find the first standard period referenced from transitions_group
        standard_periods_map = {p.id: p for p in self.periods if p.name == "Standard"}
        standard_transitions = (t for t in transitions_group.transitions if t.to in standard_periods_map)
        try:
            return next(standard_transitions)
        except StopIteration:
            raise TimezoneDefinitionInvalidForYear(f"Year {for_year} not included in periods {self.periods}")
```

This way the exception only happens in an exception situation, not whenever we're dealing with a `Daylight` transition.

File: exchangelib/properties.py (hunk 1 above)
Comment (ecederstrand) [reply exchangelib/properties.py:1927]: Ah, good catch. I agree that passing `for_year` just for the exception doesn't make sense, and the new error condition is not about the year anyway. It's not an important exception class, so we can just delete `TimezoneDefinitionInvalidForYear` entirely from this package and throw a `ValueError` here instead. Something like `ValueError(f'Transition reference {transition.to} not found in standard periods: {standard_periods_map}')`. It's just for easier debugging.

File: exchangelib/properties.py (hunk 2 above)
Comment (ecederstrand) [reply exchangelib/properties.py:1901]: A transition group will contain at most two transitions, so performance is not a concern. I prefer readability, and the for loop does that for me, at least.

File: exchangelib/properties.py (hunk 1 above)
Comment (cygnus9) [reply exchangelib/properties.py]: I just pushed a change where (amongst other changes) I do pass the year, so feel free to ignore that for now. I'll add another one with a `ValueError` exception when I get the time (later today I hope).

File: exchangelib/properties.py (hunk 2 above)
Comment (cygnus9) [reply exchangelib/properties.py]: Sure, I have no issue with that. I'll include that in my next change as well.

**Pull request merged**